### PR TITLE
Implement parsing of W3D model data in C&C3 asset streams

### PR DESCRIPTION
**Pull request by tgjones**

### CODE REVIEW

File: src/OpenSage.Game/Graphics/ModelMesh.cs [hunk 1]
```diff
@@ -71,9 +72,7 @@ public sealed class ModelMesh : DisposableBase
 
             _effect = effect;
 
-            _vertexBuffer = AddDisposable(graphicsDevice.CreateStaticBuffer(
-                vertices,
-                BufferUsage.VertexBuffer));
+            _vertexBuffer = AddDisposable(graphicsDevice.CreateVertexBuffer(vertexData));
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/ModelMesh.cs]: Indices and vertices should use the same `CreateXXX` function

Comment (tgjones) [reply src/OpenSage.Game/Graphics/ModelMesh.cs]: But `CreateStaticBuffer` takes a `T[]`, while `CreateVertexBuffer` takes a `byte[]`. I could name them both `CreateStaticBuffer`, if you'd prefer?

**Comment by BSG-75** [comment]
I tried it with RA3 and it crashed...
because it tried to load "`Data\\INI\\Default\\subsystemlegend.ini"`, and `_filesystem.GetFile()` returned `null`

With my C&C:3 it crashes as well with other kind of ini errors...

**Comment by tgjones** [comment]
Yes, it doesn’t work in the viewer yet (and rendering isn’t yet implemented anyway). I’ve just been running the ManifestFileTests while working on parsing.

**Comment by Qibbi** [comment]
For C&C3 and onward if you want support versions which aren't just the latest ones it would be important to identify an asset type not just by its type id but also it's type hash.

**Comment by Qibbi** [comment]
Also shouldn't OpenSage.Data.StreamFS.AssetReaders.AssetReader be an interface? It only has an abstract read only property and an abstract method.

**Comment by tgjones** [comment]
@qibbi interesting - does the type hash change per version? I didn’t notice that. I’m supporting multiple versions now, but only by checking the host game context, which is a bit hacky.

**Comment by tgjones** [comment]
@BSG-75 Oops - turns out the C&C3 and RA3 ini errors were because of something @paavohuhtala is working on. I've done a temporary workaround, so you can switch to C&C3 and RA3 in the viewer.

You still can't view w3d assets for those games, though, I still need to work more on that.

**Comment by Qibbi** [comment]
The change from TW to KW in W3DMesh is that they added a "GeometryPickable" bool. It doesn't change the assets size (as before it had 3 1-byte members, now it has 4, but with padding it is the same size in the end).

**Pull request merged**